### PR TITLE
[Diffusion] Support MiniMax-H3 PDD(Parallel Decoding Distillation) inference

### PR DESCRIPTION
**Pull request by IPostYellow**
Co-authored-by: @forrestl111 
## Motivation

Support offline inference with the MiniMax-H3 FL2VA and Ref2VA PDD acceleration checkpoints from [alibaba-pai/MiniMax-H3-Acc-LoRAs](https://huggingface.co/alibaba-pai/MiniMax-H3-Acc-LoRAs). These checkpoints contain backbone LoRA deltas and interval-specific output heads that require separate handling.

## Modifications

- Add tools to merge backbone LoRA weights and fuse 32 interval-specific output heads into 8 heads.
- Keep PDD heads outside the transformer checkpoint directory and preserve checkpoint indexes.
- Load fused heads through `SGLANG_DIFFUSION_MINIMAX_H3_PDD_HEADS`, with TP-aware output sharding and a shared projection path.
- Validate inference schedules against the fusion configuration. Warmup uses a prefix of the same schedule.
- Add regression tests for TP projection, checkpoint conversion, and schedule validation.

The standard inference path remains unchanged when PDD is disabled. PDD requests use `num_inference_steps=9`, corresponding to 8 denoising forwards.

## Accuracy Tests

- All 3 regression tests passed on the remote test host, covering TP projection, checkpoint conversion, and schedule validation.
- Inference was validated on NVIDIA CUDA with **TP=2, SP/Ulysses=4** and **TP=1, SP/Ulysses=8**.
- With TP2/SP4, completed **18 generations covering 9 paired comparisons**: T2VA, FL2VA, and video-reference Ref2VA, each at 5/10/15 seconds with seed 0.
- SSIM compares each PDD output against a freshly generated lossless output with identical input and seed.
- Metric: frame-mean YUV420 SSIM with Y:U:V weights of 4:1:1. Resolution and frame count were checked before comparison.
- TP1/SP8 smoke tests passed for T2VA, FL2VA, and video-reference Ref2VA using 5-second samples.

| Task | Duration | SSIM (PDD vs lossless) |
|---|---:|---:|
| T2VA | 5s | 0.7102 |
| T2VA | 10s | 0.6614 |
| T2VA | 15s | 0.7614 |
| FL2VA | 5s | 0.6300 |
| FL2VA | 10s | 0.6543 |
| FL2VA | 15s | 0.5535 |
| Ref2VA (video) | 5s | 0.8130 |
| Ref2VA (video) | 10s | 0.9788 |
| Ref2VA (video) | 15s | 0.8666 |

## Speed Tests and Profiling

Configuration: **8× NVIDIA H200, TP=2, SP/Ulysses=4, dense FlashAttention, 768p**.

Both modes use identical inputs and seed 0. Lossless uses 49 denoising forwards; PDD uses 8. Each server receives a warmup request before measurement.

The table reports server-side generation time from **`Pixel data generated successfully in ... seconds`**, excluding model loading, warmup, client polling, and video download.

| Task | Duration | Lossless (s) | PDD (s) |
|---|---:|---:|---:|
| T2VA | 5s | 41.16 | 8.03 |
| T2VA | 10s | 116.00 | 21.44 |
| T2VA | 15s | 224.78 | 40.52 |
| FL2VA | 5s | 44.61 | 8.94 |
| FL2VA | 10s | 120.21 | 22.12 |
| FL2VA | 15s | 231.89 | 41.25 |
| Ref2VA (video) | 5s | 138.78 | 27.79 |
| Ref2VA (video) | 10s | 432.25 | 79.52 |
| Ref2VA (video) | 15s | 864.08 | 153.75 |

Aggregate speedup, calculated as total lossless time divided by total PDD time:

- **Server-side generation: 5.49×**
- **Denoising: 6.05×**

These results use one measured run per case and one seed.

## Usage

Merge the adapter into a native MiniMax-H3 transformer checkpoint, then fuse the PDD heads:

```bash
python -m sglang.multimodal_gen.tools.build_minimax_h3_pdd_weights \
    <native-transformer-dir> <pdd-lora.safetensors> <output-dir>

python -m sglang.multimodal_gen.tools.fuse_minimax_h3_pdd_heads \
    <output-dir>
```

Enable the fused heads:

```bash
export SGLANG_DIFFUSION_MINIMAX_H3_PDD_HEADS=<output-dir>/pdd_fused_heads.safetensors
```

Pass `--transformer-weights-path <output-dir>/transformer` to the server and set `num_inference_steps=9` in requests. Keep `pdd_config.json` beside the fused heads; request flow shifts must match the fusion configuration.
## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :no_entry_sign: [Run #35711920226](https://github.com/sgl-project/sglang/actions/runs/35711920226)
Latest PR Test (Extra): :no_entry_sign: [Run #35711920333](https://github.com/sgl-project/sglang/actions/runs/35711920333)
Latest PR Test (AMD ROCm 10): :hourglass_flowing_sand: [Run #35711920502](https://github.com/sgl-project/sglang/actions/runs/35711920502)

### CODE REVIEW

File: test/registered/unit/diffusion/test_minimax_h3_pdd_offline.py [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+            _list_safetensors_files(str(out / "transformer")),
+            [str(out / "transformer/model.safetensors")],
+        )
+        torch.testing.assert_close(
+            load_file(str(out / "transformer/model.safetensors"))[
+                "blocks.0.attn.out_proj.weight"
+            ],
+            torch.ones(2, 3),
+        )
+        self.assertTrue((out / "pdd_fused_heads.safetensors").is_file())
```
Comment (BBuf) [inline test/registered/unit/diffusion/test_minimax_h3_pdd_offline.py]: Can you clean up the test file?

Comment (IPostYellow) [reply test/registered/unit/diffusion/test_minimax_h3_pdd_offline.py]: Done, cleaned up the tests and simplified the build script. All tests pass, and generation outputs are unchanged.

**Review by BBuf — APPROVED** [review]

**Comment by BBuf** [comment]
https://github.com/sgl-project/sglang/actions/runs/35711920226/job/106763294512?pr=40568

**Comment by BBuf** [comment]
By the way, should we update H3 recipe?

**Comment by IPostYellow** [comment]
> By the way, should we update H3 recipe?

Added a PDD section to the H3 cookbook with weight preparation, serving commands, and sampling settings.
